clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -3086,17 +3086,32 @@ static void handleWorkGroupSize(Sema &S, Decl *D, const ParsedAttr &AL) {
       return;
     ZDimExpr = ZDim.get();
 
+    // If the declaration has an [[intel::num_simd_work_items()]] attribute,
+    // check to see if the first argument of
+    // __attribute__((reqd_work_group_size)) attribute and last argument of
+    // [[cl::reqd_work_group_size()]] or [intel::reqd_work_group_size()]]
+    // attribute can be evenly divided by the [[intel::num_simd_work_items()]]
+    // attribute. GNU spelling of ReqdWorkGroupSizeAttr maps to the OpenCL
+    // semantics. First and last argument are only swapped for the atributes in
+    // SYCL and not the OpenCL ones.
     if (const auto *A = D->getAttr<SYCLIntelNumSimdWorkItemsAttr>()) {
       int64_t NumSimdWorkItems =
           A->getValue()->getIntegerConstantExpr(Ctx)->getSExtValue();
 
-      if (XDimVal.getZExtValue() % NumSimdWorkItems != 0) {
+      bool UsesOpenCLArgOrdering =
+          AL.getSyntax() == AttributeCommonInfo::AS_GNU;
+
+      unsigned WorkGroupSize = UsesOpenCLArgOrdering ? XDimVal.getZExtValue()
+                                                     : ZDimVal.getZExtValue();
+
+      if (WorkGroupSize % NumSimdWorkItems != 0) {
         S.Diag(A->getLocation(), diag::err_sycl_num_kernel_wrong_reqd_wg_size)
             << A << AL;
         S.Diag(AL.getLoc(), diag::note_conflicting_attribute);
         return;
       }
     }
+
```

Comment:
Somewhere (maybe in here?), I think we need to check the dimensionality of the kernel as well (see https://github.com/KhronosGroup/SYCL-Docs/issues/142).

Reply:
@AaronBallman, will it be a separate diagnostic ? or does it need to add with diag::err_sycl_num_kernel_wrong_reqd_wg_size?

Reply:
I think it'll like be a separate diagnostic, and we can add it in a follow-up patch.

Reply:
Created bug report here: https://github.com/intel/llvm/issues/3743